docs/flip/1777-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+
+By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.
+
+
+# Implementation
+
+A simplified example implementation:
+
+
+```python
+def promote_arrays(*xs, dtype):
+ if dtype is None:
+   dtype = jnp.result_type(*jax.tree_leaves(xs))
+ return jax.tree_map(lambda x: jnp.asarray(x, dtype), xs)
+
+Dtype = Any
+class Dense(nn.Module):
+ features: int
+ kernel_init: Callable
+ bias_init: Callable
+ dtype: Optional[Dtype] = None
+ param_dtype: Dtype = jnp.float32
+
+ @nn.compact
+ def __call__(self, x):
+   kernel = self.param("kernel", 
+                       self.kernel_init,
+                       (x.shape[-1], self.features), self.param_dtype)
+   bias = self.param("bias", self.bias_init, (self.features,), self.param_dtype)
+   x, kernel, bias = promote_arrays(x, kernel, bias, dtype=self.dtype)
+   return x @ kernel + bias
+```
+
+
+## Half precision types
+
+Some layers don’t work with half-precision types internally. For example: The normalization layers currently compute mean and variance in float32 even when a half precision type is specified to avoid numerical issues. We can replicate this behavior by calling result\_dtype with a dummy argument that has the minimum precision for the sub computation to work correctly.
+
+
+## Backward compatibility
+
+This proposal causes some layers to behave differently in cases where the dtype is not specified to a linen Module. By default, parameters are in float32. Therefore, passing in half or float32 precision inputs will cause a float32 dtype and no functional differences with current behavior.
+
+When passing complex or float64 precision, the result will no longer truncate the imaginary component or the precision. The silent truncation is problematic and has caused user complaints. Therefore, this change can be considered a bugfix.
+
+
+## Corner cases
+
+How to handle non-parameter output state on a case-by-case basis:
+
+
+**Autoregressive decoding cache**
+
+Currently, only attention implements autoregressive cache and the stored key and value mirror the dtype of the key and value passed to the layer. Forcing the cache dtype to be the same as the output dtype could result in reduced precision during cached decoding vs uncached. This seems undesirable. Decision: keep the current behavior.
+
+**Batch statistics**
```

Comment:
Above you mention "normalization layers currently compute mean and variance in float32" and here you only talk about batch norm - is that on purpose?

Reply:
I agree this seems to apply to both layer norm and batch norm. Perhaps it is useful to link to this section where you discuss the normalization layers above?

Reply:
No batchnorm is different because it also stores batch statistics which the other norm layers do not